zerostash/src/commands/mount.rs

```diff
@@ -0,0 +1,279 @@
+            file_content: Mutex::new(HashMap::new()),
+            dir_map,
+            file_attr_memory: Mutex::new(HashMap::new()),
+        })
+    }
+}
+
+const TTL: Duration = Duration::from_secs(1);
+
+pub fn file_to_fuse(file: &Arc<Entry>) -> FileAttr {
+    FileAttr {
+        size: file.size,
+        blocks: 1,
+        atime: SystemTime::now(),
+        mtime: SystemTime::now(),
+        ctime: SystemTime::now(),
+        crtime: SystemTime::UNIX_EPOCH,
+        kind: FileType::RegularFile,
+        perm: 0o444,
+        nlink: 1,
+        gid: file.unix_gid.unwrap_or(0),
+        uid: file.unix_uid.unwrap_or(0),
+        rdev: 0,
+        flags: 0,
+    }
+}
+
+fn strip_path(path: &Path) -> &Path {
+    path.strip_prefix("/").unwrap()
+}
+
+impl FilesystemMT for SimpleFs {
+    fn destroy(&self) {
+        self.destroy_tx
+            .send(())
+            .expect("Could not send signal on channel.")
+    }
+    fn getattr(&self, _req: RequestInfo, path: &Path, _fh: Option<u64>) -> ResultEntry {
+        println!("getattr = {:?}", path);
+
+        let real_path = strip_path(path);
+        let mut file_attr_memory = self.file_attr_memory.lock().unwrap();
+
+        if self.dir_map.contains_key(real_path) {
+            Ok((TTL, DIR_ATTR))
+        } else if let Some(file_attr) = file_attr_memory.get(real_path) {
+            Ok((TTL, *file_attr))
+        } else {
+            match self.file_map.get(real_path) {
+                Some(metadata) => {
+                    let fuse = file_to_fuse(metadata);
+                    file_attr_memory.insert(real_path.into(), fuse);
+                    Ok((TTL, fuse))
+                }
+                None => Err(libc::ENOENT),
+            }
+        }
+    }
+
+    fn opendir(&self, _req: RequestInfo, _path: &Path, _flags: u32) -> ResultOpen {
+        println!("(opendir: {:?} flags = {:#o})", _path, _flags);
+        Ok((0, 0))
+    }
+
+    fn readdir(&self, _req: RequestInfo, path: &Path, _fh: u64) -> ResultReaddir {
+        println!("readdir: {:?}", path);
+        Ok(self
+            .dir_map
+            .get(strip_path(path))
+            .cloned()
+            .unwrap_or_default())
+    }
+    fn open(&self, _req: RequestInfo, path: &Path, _flags: u32) -> ResultOpen {
+        println!("open: {:?} (flags = {:#x})", path, _flags);
+
+        let mut file_content = self.file_content.lock().unwrap();
+        let real_path = strip_path(path);
+
+        if file_content.get(real_path).is_none() {
+            let metadata = self.file_map.get(real_path).unwrap();
+            let mut objectreader = self.stash.storage_reader().unwrap();
+            let mut buf = vec![];
+            buf.resize(metadata.size as usize, 0);
+
+            for (start, cp) in metadata.chunks.iter() {
+                let start = *start as usize;
+                objectreader.read_chunk(cp, &mut buf[start..]).unwrap();
+            }
+
+            file_content.insert(PathBuf::from(real_path), buf);
```

Comment:
This effectively means that for every file that's accessed, we store it in memory without a cleanup mechanism. This is not great when you're thinking about large VM images, or other opaque binary data, for instance.

Instead I would push this logic into `read` and access chunks on demand without a full reconstruction in memory. This means `open` would basically be a no-op. Is there a reason why that isn't a viable approach?

Reply:
Do you mean by "access chunks on demand" that I just do `read_chunk()` on the parts I need to return in `read()`?

What if I push everything into `read()` as you recommended, store the whole data and `release()` it right after again, so we have that cleanup mechanism.

Reply:
Yes, I mean access the chunks on demand based on the positioning information in the arguments of `read()`.

Storing the whole data may not be feasible -- what if I want to restore a 128GB disk image? You'd need a lot of RAM and swap to be able to store the file in memory, not available on most consumer machines. Yet, working with ~100GB disk images may not be out of the ordinary for some, as 1TB SSD is very affordable.